<!DOCTYPE html>
<html lang="zh-cn">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>使用KinFin进行比较基因组分析 - SR-C</title>
  <meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>

<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">


<meta name="author" content="SR-C" /><meta name="description" content="安装 1 2 3 git clone https://github.com/DRL/kinfin.git cd kinfin ./install 安装程序会下载一些数据，并使用pip安装相应的模块。但若是没有管理员权限，还是使用conda来配置环境比较方便。 软件的" /><meta name="keywords" content="Genomics, Bioinformation" />






<meta name="generator" content="Hugo 0.74.3 with theme even" />


<link rel="canonical" href="https://sr-c.github.io/2019/03/23/KinFin-quickstart/" />
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">

<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<link href="/sass/main.min.8c3cbcb0324c2bb4875ceccba4007cbad4b4ac8377f33af9953c3e7684534a50.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css" integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin="anonymous">


<meta property="og:title" content="使用KinFin进行比较基因组分析" />
<meta property="og:description" content="安装 1 2 3 git clone https://github.com/DRL/kinfin.git cd kinfin ./install 安装程序会下载一些数据，并使用pip安装相应的模块。但若是没有管理员权限，还是使用conda来配置环境比较方便。 软件的" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://sr-c.github.io/2019/03/23/KinFin-quickstart/" />
<meta property="article:published_time" content="2019-03-23T18:31:47+00:00" />
<meta property="article:modified_time" content="2019-03-23T18:31:47+00:00" />
<meta itemprop="name" content="使用KinFin进行比较基因组分析">
<meta itemprop="description" content="安装 1 2 3 git clone https://github.com/DRL/kinfin.git cd kinfin ./install 安装程序会下载一些数据，并使用pip安装相应的模块。但若是没有管理员权限，还是使用conda来配置环境比较方便。 软件的">
<meta itemprop="datePublished" content="2019-03-23T18:31:47+00:00" />
<meta itemprop="dateModified" content="2019-03-23T18:31:47+00:00" />
<meta itemprop="wordCount" content="1006">



<meta itemprop="keywords" content="KinFin," />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="使用KinFin进行比较基因组分析"/>
<meta name="twitter:description" content="安装 1 2 3 git clone https://github.com/DRL/kinfin.git cd kinfin ./install 安装程序会下载一些数据，并使用pip安装相应的模块。但若是没有管理员权限，还是使用conda来配置环境比较方便。 软件的"/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->

</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">SR-C</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <a href="/">
        <li class="mobile-menu-item">主页</li>
      </a><a href="/post/">
        <li class="mobile-menu-item">归档</li>
      </a><a href="/tags/">
        <li class="mobile-menu-item">标签</li>
      </a><a href="/categories/">
        <li class="mobile-menu-item">分类</li>
      </a><a href="/search/">
        <li class="mobile-menu-item">搜索</li>
      </a>
  </ul>
</nav>
  <div class="container" id="mobile-panel">
    <header id="header" class="header">
        <div class="logo-wrapper">
  <a href="/" class="logo">SR-C</a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    <li class="menu-item">
        <a class="menu-item-link" href="/">主页</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/post/">归档</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/tags/">标签</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/categories/">分类</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/search/">搜索</a>
      </li>
  </ul>
</nav>
    </header>

    <main id="main" class="main">
      <div class="content-wrapper">
        <div id="content" class="content">
            
			<article class="post">
    
    <header class="post-header">
      <h1 class="post-title">使用KinFin进行比较基因组分析</h1>

      <div class="post-meta">
        <span class="post-time"> 2019-03-23 </span>
        
        <span id="busuanzi_container_page_pv" class="more-meta"> <span id="busuanzi_value_page_pv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> 次阅读 </span>
      </div>
    </header>

    <div class="post-toc" id="post-toc">
  <h2 class="post-toc-title">文章目录</h2>
  <div class="post-toc-content always-active">
    <nav id="TableOfContents">
  <ul>
    <li><a href="#安装">安装</a></li>
    <li><a href="#测试">测试</a></li>
    <li><a href="#进阶使用">进阶使用</a></li>
    <li><a href="#参考来源">参考来源</a></li>
  </ul>
</nav>
  </div>
</div>
    <div class="post-content">
      <!-- raw HTML omitted -->
<!-- raw HTML omitted -->
<h2 id="安装">安装</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">git clone https://github.com/DRL/kinfin.git
<span class="nb">cd</span> kinfin
./install
</code></pre></td></tr></table>
</div>
</div><p>安装程序会下载一些数据，并使用<code>pip</code>安装相应的模块。但若是没有管理员权限，还是使用<code>conda</code>来配置环境比较方便。</p>
<p>软件的<a href="https://github.com/DRL/kinfin/blob/master/requirements.txt">环境要求</a><code>python</code>模块如下</p>
<blockquote>
<p>scipy==0.19.0
matplotlib==2.0.2
docopt==0.6.2
networkx==1.11
powerlaw==1.4.1
ete3==3.0.0b35</p>
</blockquote>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">conda create -n kinfin <span class="nv">python</span><span class="o">=</span>2.7
<span class="nb">source</span> activate kinfin
<span class="c1">#matplotlib 2.0.2 与ete3 3.0.0b35的conda环境依赖存在冲突，故直接尝试安装最新版</span>
conda install <span class="nv">scipy</span><span class="o">==</span>0.19.0 matplotlib <span class="nv">docopt</span><span class="o">==</span>0.6.2 <span class="nv">networkx</span><span class="o">==</span>1.11 ete3 
<span class="c1">#powerlaw目前只能从第三方源安装</span>
conda install -c mlgill powerlaw 

./kinfin --help
</code></pre></td></tr></table>
</div>
</div><p>出现帮助信息表明软件安装完成</p>
<p>此外，还可以安装一些<a href="https://kinfin.readme.io/docs/getting-started#section-additional-useful-software">其他软件</a></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">conda install parallel

<span class="nb">cd</span> ..
wget https://github.com/gephi/gephi/releases/download/v0.9.2/gephi-0.9.2-linux.tar.gz
tar zxf gephi-0.9.2-linux.tar.gz

<span class="c1"># 使用trimAl截短多序列比对后的序列</span>
<span class="c1"># http://trimal.cgenomics.org/downloads</span>
wget https://github.com/scapella/trimal/archive/v1.4.1.tar.gz -O trimAl_v1.4.1.tar.gz
tar zxf trimAl_v1.4.1.tar.gz
<span class="nb">cd</span> trimal-1.4.1/source
make

git clone https://github.com/PatrickKueck/FASconCAT.git
</code></pre></td></tr></table>
</div>
</div><h2 id="测试">测试</h2>
<p>KinFin的软件路径中提供了测试数据<code>./test</code>，但我们还可以使用<a href="https://kinfin.readme.io/docs/running-the-example-dataset">教程</a>中的数据来练习。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span><span class="lnt">56
</span><span class="lnt">57
</span><span class="lnt">58
</span><span class="lnt">59
</span><span class="lnt">60
</span><span class="lnt">61
</span><span class="lnt">62
</span><span class="lnt">63
</span><span class="lnt">64
</span><span class="lnt">65
</span><span class="lnt">66
</span><span class="lnt">67
</span><span class="lnt">68
</span><span class="lnt">69
</span><span class="lnt">70
</span><span class="lnt">71
</span><span class="lnt">72
</span><span class="lnt">73
</span><span class="lnt">74
</span><span class="lnt">75
</span><span class="lnt">76
</span><span class="lnt">77
</span><span class="lnt">78
</span><span class="lnt">79
</span><span class="lnt">80
</span><span class="lnt">81
</span><span class="lnt">82
</span><span class="lnt">83
</span><span class="lnt">84
</span><span class="lnt">85
</span><span class="lnt">86
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash"><span class="c1"># $KINFIN/: path to KinFin installation</span>
<span class="c1"># $MAFFT/: path to MAFFT installation</span>
<span class="c1"># $TRIMAL/: path to trimal installation</span>
<span class="c1"># $FASCONCAT/: path to FASconCAT installation</span>
<span class="c1"># $RAXML/: path to RAxML installation</span>

<span class="c1"># 1. Clone KinFin manuscript repo</span>
git clone https://github.com/DRL/kinfin_manuscript
<span class="nb">cd</span> kinfin_manuscript/

<span class="c1"># 2. Basic analysis</span>
tar zxf supplementary_material/supplementary_data_1.tar.gz
<span class="nv">$KINFIN</span>/kinfin <span class="se">\
</span><span class="se"></span>	-g supplementary_data_1/Orthogroups.I1_5.txt <span class="se">\
</span><span class="se"></span>  -c supplementary_data_1/kinfin.config.basic.txt <span class="se">\
</span><span class="se"></span>  -s supplementary_data_1/kinfin.SequenceIDs.txt <span class="se">\
</span><span class="se"></span>  -o kinfin.basic 
  
<span class="c1"># 2.1 Get single-copy cluster IDs</span>
grep <span class="s1">&#39;true&#39;</span> kinfin.basic.kinfin_results/all/all.all.cluster_1to1s.txt <span class="p">|</span> <span class="se">\
</span><span class="se"></span>	cut -f1 &gt; single_copy.cluster_ids.txt

<span class="c1"># 2.2 Get protein IDs from single-copy cluster IDs</span>
<span class="nv">$KINFIN</span>/scripts/get_protein_ids_from_cluster.py <span class="se">\
</span><span class="se"></span>	-g supplementary_data_1/Orthogroups.I1_5.txt <span class="se">\
</span><span class="se"></span>  --cluster_ids single_copy.cluster_ids.txt

<span class="c1"># 2.3 Extract FASTA sequences based on protein IDs</span>
tar zxf supplementary_material/supplementary_data_2.tar.gz
cat supplementary_data_2/fastas/*.faa &gt; all.proteins.faa
parallel -j8 <span class="s1">&#39;\
</span><span class="s1">	grep --no-group-separator -A1 -wFf {} all.proteins.faa &gt; {/.}.faa\
</span><span class="s1">  &#39;</span> ::: Orthogroups.I1_5.OG*.txt
  
<span class="c1"># 2.4 Align FASTA sequences</span>
parallel -j <span class="m">8</span> <span class="s1">&#39;\
</span><span class="s1">	$MAFFT/bin/einsi {} &gt; {/.}.einsi.faa&#39;</span> ::: Orthogroup*.faa
 
<span class="c1"># 2.5 Trim alignments</span>
parallel -j <span class="m">8</span> <span class="s1">&#39;\
</span><span class="s1">	$TRIMAL/source/trimal -in {} -out {/.}.trimal.faa -automated1 \
</span><span class="s1">  &#39;</span> ::: *einsi.faa
  
<span class="c1">## ERROR: Alignment not loaded: &#34;Orthogroup.I1_5.OG*.einsi.faa&#34; Check the file&#39;s content.</span>

<span class="c1"># 2.6 Sanitise headers for FASconCAT</span>
parallel -j <span class="m">8</span> <span class="s1">&#39;\
</span><span class="s1">	cut -f1 -d&#34;.&#34; {} &gt; {/.}.sane.fas \
</span><span class="s1">  &#39;</span> ::: *trimal.faa

<span class="c1"># 2.7 Concatenate alignments</span>
<span class="nv">$FASCONCAT</span>/FASconCAT_v1.11.pl -s -p -n

<span class="c1"># 2.8 RAxML 注意耗时较长，推荐后台运行</span>
<span class="nv">$RAXML</span>/raxmlHPC-PTHREADS-SSE3 -s FcC_smatrix.phy -m PROTGAMMAGTR -T <span class="m">32</span> -n ml -N <span class="m">20</span> -p <span class="m">19</span>
<span class="nv">$RAXML</span>/raxmlHPC-PTHREADS-SSE3 -m PROTGAMMAGTR -T <span class="m">48</span> -n bs -p <span class="m">19</span> -b <span class="m">19</span> -# <span class="m">100</span> -s FcC_smatrix.phy
<span class="nv">$RAXML</span>/raxmlHPC-PTHREADS-SSE3 -m GTRCAT -p <span class="m">19</span> -f b -t RAxML_bestTree.ml -z RAxML_bootstrap.bs -n final

<span class="c1"># 3. Advanced analysis (assuming KinFin executable is in your $PATH)</span>
<span class="nv">$KINFIN</span>/kinfin <span class="se">\
</span><span class="se"></span>	-g supplementary_data_2/Orthogroups.I1_5.txt <span class="se">\
</span><span class="se"></span>  -c supplementary_data_2/kinfin.config.tree.txt <span class="se">\
</span><span class="se"></span>  -s supplementary_data_2/kinfin.SequenceIDs.txt <span class="se">\
</span><span class="se"></span>  -o kinfin.advanced <span class="se">\
</span><span class="se"></span>  -p supplementary_data_2/kinfin.SpeciesIDs.txt <span class="se">\
</span><span class="se"></span>  -a supplementary_data_2/fastas/ <span class="se">\
</span><span class="se"></span>  -t supplementary_data_2/kinfin.tree.nwk <span class="se">\
</span><span class="se"></span>  -f supplementary_data_2/kinfin.functional_annotation.txt
  
<span class="c1"># 4. Infer representative functional annotation (all clusters)</span>
<span class="nv">$KINFIN</span>/scripts/filter_functional_annotation_of_clusters.py all <span class="se">\
</span><span class="se"></span>	-f kinfin.advanced.kinfin_results/cluster_metrics_domains.IPR.txt <span class="se">\
</span><span class="se"></span>  -c kinfin.advanced.kinfin_results/cluster_counts_by_taxon.txt <span class="se">\
</span><span class="se"></span>  -x 0.75 <span class="se">\
</span><span class="se"></span>  -p 0.75 <span class="se">\
</span><span class="se"></span>  -o kinfin.IPR

<span class="c1"># 4. Infer representative functional annotation (synapomorphic clusters)</span>
<span class="nv">$KINFIN</span>/scripts/filter_functional_annotation_of_clusters.py synapo <span class="se">\
</span><span class="se"></span>	-f kinfin.advanced.kinfin_results/cluster_metrics_domains.IPR.txt <span class="se">\
</span><span class="se"></span>  -c kinfin.advanced.kinfin_results/cluster_counts_by_taxon.txt <span class="se">\
</span><span class="se"></span>  -t kinfin.advanced.kinfin_results/tree/tree.cluster_metrics.txt <span class="se">\
</span><span class="se"></span>  -n 0.75 <span class="se">\
</span><span class="se"></span>  -x 0.75 <span class="se">\
</span><span class="se"></span>  -p 0.75 <span class="se">\
</span><span class="se"></span>  -o kinfin.IPR
</code></pre></td></tr></table>
</div>
</div><h2 id="进阶使用">进阶使用</h2>
<p>综合功能注释信息。</p>
<p>根据其提供的说明书<code>KinFin_v1.supplementary_methods.pdf</code>。综合功能注释的方法如下：</p>
<p>功能注释</p>
<ol>
<li>短于30 个氨基酸或其中还有提前终止密码子的序列被过滤<code>filter_fastas_before_clustering.py</code></li>
<li>非最长的序列被过<code>filter_isoforms_based_on_gff3.py</code></li>
<li>所有序列通过<code>InterProScan v5.22-61.0</code>的<code>Pfam-30.0</code>与<code> SignalP-EUK-4.1</code>注释，并通过<code>iprs_to_table.py</code>转化为可识别的格式。</li>
</ol>
<h2 id="参考来源">参考来源</h2>
<p><a href="https://kinfin.readme.io/docs/running-the-example-dataset">https://kinfin.readme.io/docs/running-the-example-dataset</a></p>
<!-- raw HTML omitted -->

    </div>

    <div class="post-copyright">
  <p class="copyright-item">
    <span class="item-title">文章作者</span>
    <span class="item-content">SR-C</span>
  </p>
  <p class="copyright-item">
    <span class="item-title">上次更新</span>
    <span class="item-content">
        2019-03-23
        
    </span>
  </p>
  
  <p class="copyright-item">
    <span class="item-title">许可协议</span>
    <span class="item-content"><a rel="license noopener" href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank">CC BY-NC-ND 4.0</a></span>
  </p>
</div>
<footer class="post-footer">
      <div class="post-tags">
          <a href="/tags/KinFin/">KinFin</a>
          </div>
      <nav class="post-nav">
        <a class="prev" href="/2019/03/31/debug-EOF-within-quoted-string/">
            <i class="iconfont icon-left"></i>
            <span class="prev-text nav-default">【debug】EOF within quoted string</span>
            <span class="prev-text nav-mobile">上一篇</span>
          </a>
        <a class="next" href="/2019/03/19/LACHESIS-install/">
            <span class="next-text nav-default">LACHESIS安装</span>
            <span class="next-text nav-mobile">下一篇</span>
            <i class="iconfont icon-right"></i>
          </a>
      </nav>
    </footer>
  </article>
        </div>
        

  

  
    <script src="https://utteranc.es/client.js"
            repo="sr-c/blog-comment"
            issue-term="pathname"
            theme="github-light"
            crossorigin="anonymous"
            async>
    </script>
    <noscript>Please enable JavaScript to view the <a href="https://github.com/utterance">comments powered by utterances.</a></noscript>

      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="social-links">
      <a href="https://sr-c.github.io" class="iconfont icon-github" title="github"></a>
  <a href="https://sr-c.github.io/index.xml" type="application/rss+xml" class="iconfont icon-rss" title="rss"></a>
</div>

<div class="copyright">
  <span class="power-by">
    由 <a class="hexo-link" href="https://gohugo.io">Hugo</a> 强力驱动
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    主题 - 
    <a class="theme-link" href="https://github.com/olOwOlo/hugo-theme-even">Even</a>
  </span>

  <div class="busuanzi-footer">
    <span id="busuanzi_container_site_pv"> 本站总访问量 <span id="busuanzi_value_site_pv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> 次 </span>
      <span class="division">|</span>
    <span id="busuanzi_container_site_uv"> 本站总访客数 <span id="busuanzi_value_site_uv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> 人 </span>
  </div>

  <span class="copyright-year">
    &copy; 
    2017 - 
    2021
    <span class="heart">
      <i class="iconfont icon-heart"></i>
    </span>
    <span class="author">SR-C</span>
  </span>
</div>
      
    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont icon-up"></i>
    </div>
  </div>
  
  <script src="https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js" integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js" integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin="anonymous"></script>



<script type="text/javascript" src="/js/main.min.d7b7ada643c9c1a983026e177f141f7363b4640d619caf01d8831a6718cd44ea.js"></script>









</body>
</html>
