<!DOCTYPE html>
<html lang="zh-cn">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>使用jcvi进行MCscan可视化 - SR-C</title>
  <meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>

<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">


<meta name="author" content="SR-C" /><meta name="description" content="安装 软件推荐使用pip安装，但是作为普通用户，还是选择使用conda安装 1 2 3 4 5 conda create -n jcvi python=2.7 source activate jcvi # 安装前需要确认存在bioconda和r c" /><meta name="keywords" content="Genomics, Bioinformation" />






<meta name="generator" content="Hugo 0.74.3 with theme even" />


<link rel="canonical" href="https://sr-c.github.io/2019/01/11/jcvi-MCscan/" />
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">

<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<link href="/sass/main.min.8c3cbcb0324c2bb4875ceccba4007cbad4b4ac8377f33af9953c3e7684534a50.css" rel="stylesheet">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css" integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin="anonymous">


<meta property="og:title" content="使用jcvi进行MCscan可视化" />
<meta property="og:description" content="安装 软件推荐使用pip安装，但是作为普通用户，还是选择使用conda安装 1 2 3 4 5 conda create -n jcvi python=2.7 source activate jcvi # 安装前需要确认存在bioconda和r c" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://sr-c.github.io/2019/01/11/jcvi-MCscan/" />
<meta property="article:published_time" content="2019-01-11T19:19:34+00:00" />
<meta property="article:modified_time" content="2019-01-11T19:19:34+00:00" />
<meta itemprop="name" content="使用jcvi进行MCscan可视化">
<meta itemprop="description" content="安装 软件推荐使用pip安装，但是作为普通用户，还是选择使用conda安装 1 2 3 4 5 conda create -n jcvi python=2.7 source activate jcvi # 安装前需要确认存在bioconda和r c">
<meta itemprop="datePublished" content="2019-01-11T19:19:34+00:00" />
<meta itemprop="dateModified" content="2019-01-11T19:19:34+00:00" />
<meta itemprop="wordCount" content="1506">



<meta itemprop="keywords" content="jcvi,MCscan," />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="使用jcvi进行MCscan可视化"/>
<meta name="twitter:description" content="安装 软件推荐使用pip安装，但是作为普通用户，还是选择使用conda安装 1 2 3 4 5 conda create -n jcvi python=2.7 source activate jcvi # 安装前需要确认存在bioconda和r c"/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->

</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">SR-C</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <a href="/">
        <li class="mobile-menu-item">主页</li>
      </a><a href="/post/">
        <li class="mobile-menu-item">归档</li>
      </a><a href="/tags/">
        <li class="mobile-menu-item">标签</li>
      </a><a href="/categories/">
        <li class="mobile-menu-item">分类</li>
      </a><a href="/search/">
        <li class="mobile-menu-item">搜索</li>
      </a>
  </ul>
</nav>
  <div class="container" id="mobile-panel">
    <header id="header" class="header">
        <div class="logo-wrapper">
  <a href="/" class="logo">SR-C</a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    <li class="menu-item">
        <a class="menu-item-link" href="/">主页</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/post/">归档</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/tags/">标签</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/categories/">分类</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/search/">搜索</a>
      </li>
  </ul>
</nav>
    </header>

    <main id="main" class="main">
      <div class="content-wrapper">
        <div id="content" class="content">
            
			<article class="post">
    
    <header class="post-header">
      <h1 class="post-title">使用jcvi进行MCscan可视化</h1>

      <div class="post-meta">
        <span class="post-time"> 2019-01-11 </span>
        
        <span id="busuanzi_container_page_pv" class="more-meta"> <span id="busuanzi_value_page_pv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> 次阅读 </span>
      </div>
    </header>

    <div class="post-toc" id="post-toc">
  <h2 class="post-toc-title">文章目录</h2>
  <div class="post-toc-content always-active">
    <nav id="TableOfContents">
  <ul>
    <li><a href="#安装">安装</a></li>
    <li><a href="#准备数据">准备数据</a></li>
    <li><a href="#共线性分析">共线性分析</a></li>
    <li><a href="#可视化">可视化</a>
      <ul>
        <li><a href="#点阵图">点阵图</a></li>
        <li><a href="#共线型图">共线型图</a></li>
        <li><a href="#局部可视化">局部可视化</a></li>
      </ul>
    </li>
    <li><a href="#参考来源">参考来源</a></li>
  </ul>
</nav>
  </div>
</div>
    <div class="post-content">
      <!-- raw HTML omitted -->
<!-- raw HTML omitted -->
<h2 id="安装">安装</h2>
<p>软件推荐使用<code>pip</code>安装，但是作为普通用户，还是选择使用conda安装</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">conda create -n jcvi <span class="nv">python</span><span class="o">=</span>2.7
<span class="nb">source</span> activate jcvi
<span class="c1"># 安装前需要确认存在bioconda和r channel，安装前先更新conda</span>
conda --add-channel r
conda install jcvi last
</code></pre></td></tr></table>
</div>
</div><p>后续还可能会由于缺少<code>scipy</code>在绘图时报错<code>ImportError: No module named scipy.spatial</code>，因此可提前安装</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash"><span class="c1"># 安装scipy</span>
<span class="c1">#conda-forge scipy-1.2.0</span>
conda install scipy
</code></pre></td></tr></table>
</div>
</div><h2 id="准备数据">准备数据</h2>
<p><code>jcvi</code>支持直接从<a href="https://phytozome.jgi.doe.gov/pz/portal.html">phytozome</a>下载数据</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.apps.fetch phytozome
</code></pre></td></tr></table>
</div>
</div><p>按照名称查看</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.apps.fetch phytozome
...
         Acoerulea               Alyrata             Athaliana
       Bdistachyon                 Brapa           Cclementina
           Cpapaya          Creinhardtii              Crubella
          Csativus             Csinensis Csubellipsoidea_C-169
          Egrandis                Fvesca                  Gmax
        Graimondii        Lusitatissimum            Mdomestica
        Mesculenta             Mguttatus     Mpusilla_CCMP1545
   Mpusilla_RCC299           Mtruncatula          Olucimarinus
           Osativa               Ppatens              Ppersica
      Ptrichocarpa             Pvirgatum             Pvulgaris
         Rcommunis              Sbicolor              Sitalica
     Slycopersicum       Smoellendorffii            Stuberosum
            Tcacao            Thalophila              Vcarteri
         Vvinifera                 Zmays         early_release
</code></pre></td></tr></table>
</div>
</div><p>常见的植物物种都可以直接下载</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash"><span class="c1"># download sequences and coordinates of grape and peach</span>
python -m jcvi.apps.fetch phytozome Vvinifera,Ppersica
<span class="c1"># convert the GFF to BED file</span>
python -m jcvi.formats.gff bed --type<span class="o">=</span>mRNA --key<span class="o">=</span>Name Vvinifera_145_gene.gff3.gz -o grape.bed
python -m jcvi.formats.gff bed --type<span class="o">=</span>mRNA --key<span class="o">=</span>Name Ppersica_139_gene.gff3.gz -o peach.bed
<span class="c1"># clean headers to remove description fiedls from Phytozome FASTA files.</span>
python -m jcvi.formats.fasta format --sep<span class="o">=</span><span class="s2">&#34;|&#34;</span> Vvinifera_145_cds.fa.gz grape.cds
python -m jcvi.formats.fasta format --sep<span class="o">=</span><span class="s2">&#34;|&#34;</span> Ppersica_139_cds.fa.gz peach.cds
</code></pre></td></tr></table>
</div>
</div><h2 id="共线性分析">共线性分析</h2>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">jcvi</span><span class="o">.</span><span class="n">compara</span><span class="o">.</span><span class="n">catalog</span> <span class="n">ortholog</span> <span class="n">grape</span> <span class="n">peach</span>
<span class="n">ls</span> <span class="n">grape</span><span class="o">.</span><span class="n">peach</span><span class="o">.*</span>
<span class="n">grape</span><span class="o">.</span><span class="n">peach</span><span class="o">.</span><span class="n">lifted</span><span class="o">.</span><span class="n">anchors</span>  <span class="n">grape</span><span class="o">.</span><span class="n">peach</span><span class="o">.</span><span class="n">anchors</span>  <span class="n">grape</span><span class="o">.</span><span class="n">peach</span><span class="o">.</span><span class="n">last</span><span class="o">.</span><span class="n">filtered</span> <span class="n">grape</span><span class="o">.</span><span class="n">peach</span><span class="o">.</span><span class="n">last</span> <span class="n">grape</span><span class="o">.</span><span class="n">peach</span><span class="o">.</span><span class="n">pdf</span>
</code></pre></td></tr></table>
</div>
</div><p>程序默认按照可用线程，调用全部进行LAST比对，速度很快。</p>
<h2 id="可视化">可视化</h2>
<h3 id="点阵图">点阵图</h3>
<p>目前版本<code>0.8.12</code>在共线性分析后，自动绘制点阵图，手动运行方式如下</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.graphics.dotplot grape.peach.anchors
</code></pre></td></tr></table>
</div>
</div><p>在两者的点阵图中，不论水平方向或者竖直方向查看，都可以发现不超过3次的共线性区域。这说明，🍇和🍑之间发生了基因组三倍体化的事件，使得出现了这样3:3的模式。</p>
<p>如果仔细观察，还可以发现3个共线性区域中常常有一个信号更强，对应着两个基因组之间的直系同源区域。如果我们只想要得到这些1:1直系同源的区域呢？我们只需要重复之前的比对，同时加上选项<code>--cscore=.99</code>即可。C-score是由LAST比对区域到BLAST比对区域的比值确定。(C-score is defined by the ratio of LAST hit to the best BLAST hits to either the query and hit. )0.99的C-score阈值有效地过滤LAST比对结果，从而得到最佳相互比对结果(reciprocal best hit, RBH)。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">rm grape.peach.last.filtered 
python -m jcvi.compara.catalog ortholog grape peach --cscore<span class="o">=</span>.99
python -m jcvi.graphics.dotplot grape.peach.anchors
</code></pre></td></tr></table>
</div>
</div><p><img src="https://camo.githubusercontent.com/222430d386faca2179528563cde9d8339bc8ad72/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f33326d6d6f306b66687478356238672f67726170652e70656163682e6373636f72652e39392e706e673f7261773d31" alt="Grape-peach-synteny-.99"></p>
<p>还可以手动查看synteny depth分布</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.compara.synteny depth --histogram grape.peach.anchors
</code></pre></td></tr></table>
</div>
</div><p><img src="https://camo.githubusercontent.com/d82b2f2ed6b675ab090564d74d9b78ac2b6e21c9/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f686878326474727972756d3267796f2f67726170652e70656163682e64657074682e706e673f7261773d31" alt="Grape-peach-depth"></p>
<h3 id="共线型图">共线型图</h3>
<p>同样由<code>grape.peach.anchors</code>生成，除了BED和共线性文件，还需要手动设置两个文件用于绘图</p>
<ol>
<li>
<p><code>seqids</code>设定绘制哪些染色体，一般会去掉较小和未定位的scaffolds，下方设定了19个🍇的染色体与8个🍑的染色体</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-fallback" data-lang="fallback">chr1,chr2,chr3,chr4,chr5,chr6,chr7,chr8,chr9,chr10,chr11,chr12,chr13,chr14,chr15,chr16,chr17,chr18,chr19
scaffold_1,scaffold_2,scaffold_3,scaffold_4,scaffold_5,scaffold_6,scaffold_7,scaffold_8
</code></pre></td></tr></table>
</div>
</div></li>
<li>
<p><code>layout</code>设定画板上如何绘制。整体上canvas绘图的x,y轴的定位在0-1之间，前三行设定了染色体轨道的位置，循环，颜色，标签，垂直排列以及对应的BED文件。下方的track0就是🍇，track1就是🍑。下一节就是设定在哪些tracks之间绘制共线性关系。<code>e,0,1</code>代表着绘制track0与track1之间的共线性，使用<code>grape.peach.anchors.simple</code>文件中的详细信息。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-fallback" data-lang="fallback"># y, xstart, xend, rotation, color, label, va,  bed
 .6,     .1,    .8,       0,      , Grape, top, grape.bed
 .4,     .1,    .8,       0,      , Peach, top, peach.bed
# edges
e, 0, 1, grape.peach.anchors.simple
</code></pre></td></tr></table>
</div>
</div></li>
</ol>
<p>注意，这两个设定文件末尾不能够存在多余的空行，否则会报错<a href="https://github.com/tanghaibao/jcvi/issues/56">#56</a></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">LayoutLine<span class="s1">&#39; object has no attribute &#39;</span>color<span class="err">&#39;</span>
</code></pre></td></tr></table>
</div>
</div><ol start="3">
<li>从<code>.anchors</code>文件中产生更简要的<code>.simple</code>文件</li>
</ol>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.compara.synteny screen --minspan<span class="o">=</span><span class="m">30</span> --simple grape.peach.anchors grape.peach.anchors.new 
</code></pre></td></tr></table>
</div>
</div><p>绘图</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.graphics.karyotype seqids layout
</code></pre></td></tr></table>
</div>
</div><p><img src="https://camo.githubusercontent.com/b43d5bd13f2abaf1255fac49d6fe8ba5b3f86ec5/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f35316b386a756a797a616765336f612f67726170652e70656163682e6b6172796f747970652e706e673f7261773d31" alt="Grape-peach-karyotype"></p>
<h3 id="局部可视化">局部可视化</h3>
<p>首先从基因水平提取匹配的区块</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">python -m jcvi.compara.synteny mcscan grape.bed grape.peach.lifted.anchors --iter<span class="o">=</span><span class="m">1</span> -o grape.peach.i1.blocks
</code></pre></td></tr></table>
</div>
</div><p>参数<code>--iter=1</code>表明从每个🍇区域中提取1个最佳匹配区域。若将<code>--iter</code>设置为2，那么每个🍇区域就会有2个对应的🍑区域。这一点对于基因组重复区域的绘制很有帮助。</p>
<p>目前<code>grape.peach.i1.blocks</code>中包含有很多局部区域，我们可以手动选择需要绘制的部分。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">head -50 grape.peach.i1.blocks &gt; blocks
</code></pre></td></tr></table>
</div>
</div><p>最终，我们仍然需要一个<code>layout</code>文件进行绘图，<code>blocks.layout</code>文件如下</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-fallback" data-lang="fallback"># x,   y, rotation,   ha,     va,   color, ratio,            label
0.5, 0.6,        0, left, center,       m,     1,       grape Chr1
0.5, 0.4,        0, left, center, #fc8d62,     1, peach scaffold_1
# edges
e, 0, 1
</code></pre></td></tr></table>
</div>
</div><p>然后，我们就可以进行绘制了</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-bash" data-lang="bash">cat grape.bed peach.bed &gt; grape_peach.bed
python -m jcvi.graphics.synteny blocks grape_peach.bed blocks.layout
</code></pre></td></tr></table>
</div>
</div><p><img src="https://camo.githubusercontent.com/be5a68f92e48eaf6173442ce8543cefd89445e2b/68747470733a2f2f7777772e64726f70626f782e636f6d2f732f6978326d736464736e6477653276652f67726170652e70656163682e626c6f636b732e706e673f7261773d31" alt="Grape-peach-blocks"></p>
<h2 id="参考来源">参考来源</h2>
<p><a href="https://github.com/tanghaibao/jcvi/wiki/MCscan-%28Python-version%29">https://github.com/tanghaibao/jcvi/wiki/MCscan-%28Python-version%29</a></p>
<p><a href="https://www.jianshu.com/p/39448b970287">https://www.jianshu.com/p/39448b970287</a></p>
<!-- raw HTML omitted -->

    </div>

    <div class="post-copyright">
  <p class="copyright-item">
    <span class="item-title">文章作者</span>
    <span class="item-content">SR-C</span>
  </p>
  <p class="copyright-item">
    <span class="item-title">上次更新</span>
    <span class="item-content">
        2019-01-11
        
    </span>
  </p>
  
  <p class="copyright-item">
    <span class="item-title">许可协议</span>
    <span class="item-content"><a rel="license noopener" href="https://creativecommons.org/licenses/by-nc-nd/4.0/" target="_blank">CC BY-NC-ND 4.0</a></span>
  </p>
</div>
<footer class="post-footer">
      <div class="post-tags">
          <a href="/tags/jcvi/">jcvi</a>
          <a href="/tags/MCscan/">MCscan</a>
          </div>
      <nav class="post-nav">
        <a class="prev" href="/2019/01/12/R4ds-exercises-2/">
            <i class="iconfont icon-left"></i>
            <span class="prev-text nav-default">R4ds-exercises-2</span>
            <span class="prev-text nav-mobile">上一篇</span>
          </a>
        <a class="next" href="/2019/01/08/R4ds-exercises-1/">
            <span class="next-text nav-default">R4ds-exercises-1</span>
            <span class="next-text nav-mobile">下一篇</span>
            <i class="iconfont icon-right"></i>
          </a>
      </nav>
    </footer>
  </article>
        </div>
        

  

  
    <script src="https://utteranc.es/client.js"
            repo="sr-c/blog-comment"
            issue-term="pathname"
            theme="github-light"
            crossorigin="anonymous"
            async>
    </script>
    <noscript>Please enable JavaScript to view the <a href="https://github.com/utterance">comments powered by utterances.</a></noscript>

      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="social-links">
      <a href="https://sr-c.github.io" class="iconfont icon-github" title="github"></a>
  <a href="https://sr-c.github.io/index.xml" type="application/rss+xml" class="iconfont icon-rss" title="rss"></a>
</div>

<div class="copyright">
  <span class="power-by">
    由 <a class="hexo-link" href="https://gohugo.io">Hugo</a> 强力驱动
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    主题 - 
    <a class="theme-link" href="https://github.com/olOwOlo/hugo-theme-even">Even</a>
  </span>

  <div class="busuanzi-footer">
    <span id="busuanzi_container_site_pv"> 本站总访问量 <span id="busuanzi_value_site_pv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> 次 </span>
      <span class="division">|</span>
    <span id="busuanzi_container_site_uv"> 本站总访客数 <span id="busuanzi_value_site_uv"><img src="/img/spinner.svg" alt="spinner.svg"/></span> 人 </span>
  </div>

  <span class="copyright-year">
    &copy; 
    2017 - 
    2021
    <span class="heart">
      <i class="iconfont icon-heart"></i>
    </span>
    <span class="author">SR-C</span>
  </span>
</div>
      
    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont icon-up"></i>
    </div>
  </div>
  
  <script src="https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js" integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js" integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin="anonymous"></script>



<script type="text/javascript" src="/js/main.min.d7b7ada643c9c1a983026e177f141f7363b4640d619caf01d8831a6718cd44ea.js"></script>









</body>
</html>
